<a href="https://colab.research.google.com/github/priyadharsh73/airbnb_eda/blob/main/airbnb_eda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# In-depth Exploratory Data Analysis of Airbnb Dataset

## Introduction
This notebook provides an in-depth exploratory data analysis (EDA) of the Airbnb dataset. The objective is to uncover insights and patterns in the data to better understand the factors influencing Airbnb listings.

### Airbnb Dataset Column Descriptions

| Column Name                      | Description                                                                 | Data Type   |
|----------------------------------|-----------------------------------------------------------------------------|-------------|
| id                               | Unique identifier for the listing                                           | int64       |
| name                             | Name of the listing                                                         | object      |
| host_id                          | Unique identifier for the host                                              | int64       |
| host_identity_verified           | Whether the host's identity is verified                                     | object      |
| host_name                        | Name of the host                                                            | object      |
| neighbourhood_group              | Grouping of neighbourhoods                                                  | object      |
| neighbourhood                    | Name of the neighbourhood                                                   | object      |
| lat                              | Latitude of the listing                                                     | float64     |
| long                             | Longitude of the listing                                                    | float64     |
| country                          | Country where the listing is located                                        | object      |
| country_code                     | Country code                                                                | object      |
| instant_bookable                 | Whether the listing is instantly bookable                                   | object      |
| cancellation_policy              | Cancellation policy of the listing                                          | object      |
| room_type                        | Type of room offered                                                        | object      |
| construction_year                | Year the property was constructed                                           | float64     |
| price                            | Price per night                                                             | float64     |
| service_fee                      | Service fee charged                                                         | float64     |
| minimum_nights                   | Minimum number of nights required to book                                   | float64     |
| number_of_reviews                | Total number of reviews received                                            | float64     |
| last_review                      | Date of the last review                                                     | datetime64  |
| reviews_per_month                | Average number of reviews per month                                         | float64     |
| review_rate_number               | Average rating of the listing                                               | float64     |
| calculated_host_listings_count   | Total number of listings by the host                                        | float64     |
| availability_365                 | Number of days the listing is available in a year                           | float64     |
| house_rules                      | House rules for the listing                                                 | object      |
| license                          | License number of the listing (if applicable)                               | object      |

## Data Loading and Preparation

### Import Libraries
First, import the necessary libraries for data manipulation, visualization, and mapping.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import folium
from folium.plugins import MarkerCluster
import zipfile
import requests
from io import BytesIO

### Download and Extract ZIP File from GitHub

This step involves downloading the ZIP file containing the Airbnb dataset from GitHub and extracting the CSV file from it. The `requests` library is used to download the file, and the `zipfile` and `io` libraries are used to handle the extraction. The extracted CSV file is then read into a pandas DataFrame for further analysis.


In [20]:
# URL of the ZIP file on GitHub (direct download link)
url = 'https://github.com/priyadharsh73/airbnb_eda/raw/main/airbnb_dataset.zip'
filename = 'Airbnb_Open_Data.csv'

# Download the ZIP file
response = requests.get(url)
zip_file = BytesIO(response.content)

# Check if the response is valid
if response.status_code == 200:
    try:
        # Extract the ZIP file
        with zipfile.ZipFile(zip_file, 'r') as z:
            # List all files in the ZIP
            print(z.namelist())

            # Extract and read the CSV file
            with z.open(filename) as f:
                df = pd.read_csv(f, low_memory=False)

        # Display the first few rows of the DataFrame
    except zipfile.BadZipFile:
        print("Error: The file is not a valid ZIP file.")
else:
    print(f"Error: Failed to download the file. Status code: {response.status_code}")


['Airbnb_Open_Data.csv']


###Clean Column Names
- Use the `str.replace()` method to replace spaces with underscores.
- Use the `str.lower()` method to convert column names to lowercase.

Here’s the code to clean the column names:

In [21]:
# Clean column names by replacing spaces with underscores and converting to lowercase
df.columns = df.columns.str.replace(' ', '_').str.lower()

# Check the cleaned column names
print("Cleaned column names:")
print(df.columns)

Cleaned column names:
Index(['id', 'name', 'host_id', 'host_identity_verified', 'host_name',
       'neighbourhood_group', 'neighbourhood', 'lat', 'long', 'country',
       'country_code', 'instant_bookable', 'cancellation_policy', 'room_type',
       'construction_year', 'price', 'service_fee', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'review_rate_number', 'calculated_host_listings_count',
       'availability_365', 'house_rules', 'license'],
      dtype='object')


In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   name                            102349 non-null  object 
 2   host_id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host_name                       102193 non-null  object 
 5   neighbourhood_group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country_code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

### Misinterpreted Columns

Based on the `info()` function output, the following columns have been misinterpreted:

- **price**: Interpreted as `object`, should be `float`.
- **service_fee**: Interpreted as `object`, should be `float`.
- **minimum_nights**: Interpreted as `float64`, should be `int`.
- **number_of_reviews**: Interpreted as `float64`, should be `int`.
- **last_review**: Interpreted as `object`, should be `date`.
- **construction_year**: Interpreted as `float64`, should be `int`.

### Code to Correct Data Types

In [23]:
# Convert price and service_fee columns to numeric after removing the dollar sign and commas
df['price'] = df['price'].replace('[\$,]', '', regex=True).astype(float)
df['service_fee'] = df['service_fee'].replace('[\$,]', '', regex=True).astype(float)

# Convert minimum_nights and number_of_reviews to integers, ignoring rows with missing values
df['minimum_nights'] = pd.to_numeric(df['minimum_nights'], errors='coerce').fillna(0).astype(int)
df['number_of_reviews'] = pd.to_numeric(df['number_of_reviews'], errors='coerce').fillna(0).astype(int)

# Convert last_review to datetime
df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')

# Convert construction_year to integer
df['construction_year'] = df['construction_year'].astype('Int64')


The following columns had their data types corrected:
- `price` and `service_fee` were converted to numeric data types after removing the dollar signs and commas.
- `minimum_nights` and `number_of_reviews` were converted to integers, with missing values filled with 0.
- `last_review` was converted to datetime.
- `construction_year` was converted to integer.

In [24]:
# Summary of missing values
missing_values_summary = df.isna().sum()

# Display the summary of missing values
missing_values_summary

id                                     0
name                                 250
host_id                                0
host_identity_verified               289
host_name                            406
neighbourhood_group                   29
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country_code                         131
instant_bookable                     105
cancellation_policy                   76
room_type                              0
construction_year                    214
price                                247
service_fee                          273
minimum_nights                         0
number_of_reviews                      0
last_review                        15893
reviews_per_month                  15879
review_rate_number                   326
calculated_host_listings_count       319
availability_365                     448
house_rules                        52131
license                           102597
dtype: int64

Insights on Missing Values
The dataset contains several columns with missing values. Here is a grouped summary of the missing values:

- **Identification and Location**:
  - `id`, `host_id`, `room_type`, `minimum_nights`, and `number_of_reviews` have no missing values.
  - `name`, `host_identity_verified`, `host_name`, `neighbourhood_group`, `neighbourhood`, `lat`, `long`, `country`, and `country_code` have a small to moderate number of missing values.

- **Booking and Policy**:
  - `instant_bookable`, `cancellation_policy`, `price`, and `service_fee` have a moderate number of missing values.
  - `construction_year` has some missing values.

- **Reviews and Availability**:
  - `last_review` and `reviews_per_month` have a significant number of missing values.
  - `review_rate_number`, `calculated_host_listings_count`, and `availability_365` have a moderate number of missing values.

- **Additional Information**:
  - `house_rules` has a large number of missing values.
  - `license` is almost entirely missing.

Handling these missing values appropriately is crucial for ensuring the quality and reliability of the analysis.

#### Dropping Insignificant Columns
- Dropped the columns `license` and `house_rules` due to a high percentage of missing values and their insignificance for the analysis.

In [25]:
# Drop insignificant columns
df.drop(columns=['license', 'house_rules'], inplace=True)

> After dropping these columns, the dataset is now more streamlined for further analysis.

#### Finding and Deleting Duplicates
Duplicates in the dataset were identified and removed based on the following important columns: `id`, `name`, `host_id`, `lat`, `long`, and `price`. This ensures that each listing is unique and prevents any potential bias or redundancy in the analysis.


In [26]:
# Count duplicates based on important columns
duplicate_count = df.duplicated(subset=['id', 'name', 'host_id', 'lat', 'long', 'price']).sum()

# Display the count of duplicates
print(f"Number of duplicate rows: {duplicate_count}")

# Drop duplicates based on important columns
df.drop_duplicates(subset=['id', 'name', 'host_id', 'lat', 'long', 'price'], keep='first', inplace=True)

# Summary of the dataset after removing duplicates
df.info()

Number of duplicate rows: 541
<class 'pandas.core.frame.DataFrame'>
Index: 102058 entries, 0 to 102057
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   id                              102058 non-null  int64         
 1   name                            101808 non-null  object        
 2   host_id                         102058 non-null  int64         
 3   host_identity_verified          101769 non-null  object        
 4   host_name                       101654 non-null  object        
 5   neighbourhood_group             102029 non-null  object        
 6   neighbourhood                   102042 non-null  object        
 7   lat                             102050 non-null  float64       
 8   long                            102050 non-null  float64       
 9   country                         101526 non-null  object        
 10  country_code                   

> The dataset contains a certain number of duplicate rows based on the important columns `id`, `name`, `host_id`, `lat`, `long`, and `price`.

After identifying the duplicates, they were removed to ensure that each listing is unique. This step helps in preventing any potential bias or redundancy in the analysis.


### Handling Missing Values
Based on the nature of each column, the following strategies were applied to handle missing values:

- **String Columns**: Filled with a placeholder "Unknown".
- **Categorical Columns**: Filled with the mode (most frequent value).
- **Numerical Columns**: Converted to numeric with `pd.to_numeric` and filled with the median value.
- **Date Columns**: Forward fill to maintain temporal sequence.
- **Critical Columns (lat, long)**: Rows with missing values were dropped.


In [29]:
# Fill missing values for string columns with a placeholder
df['name'].fillna('Unknown', inplace=True)
df['host_name'].fillna('Unknown', inplace=True)

# Fill missing values for categorical columns with the mode
df['host_identity_verified'].fillna(df['host_identity_verified'].mode()[0], inplace=True)
df['neighbourhood_group'].fillna(df['neighbourhood_group'].mode()[0], inplace=True)
df['neighbourhood'].fillna(df['neighbourhood'].mode()[0], inplace=True)
df['country'].fillna(df['country'].mode()[0], inplace=True)
df['country_code'].fillna(df['country_code'].mode()[0], inplace=True)
df['instant_bookable'].fillna(df['instant_bookable'].mode()[0], inplace=True)
df['cancellation_policy'].fillna(df['cancellation_policy'].mode()[0], inplace=True)

# Drop rows with missing values for lat and long
df.dropna(subset=['lat', 'long'], inplace=True)

# Fill missing values for numerical columns with the median
df['construction_year'] = pd.to_numeric(df['construction_year'], errors='coerce').fillna(df['construction_year'].median()).astype('Int64')
df['price'] = pd.to_numeric(df['price'], errors='coerce').fillna(df['price'].median())
df['service_fee'] = pd.to_numeric(df['service_fee'], errors='coerce').fillna(df['service_fee'].median())
df['reviews_per_month'] = pd.to_numeric(df['reviews_per_month'], errors='coerce').fillna(df['reviews_per_month'].median())
df['review_rate_number'] = pd.to_numeric(df['review_rate_number'], errors='coerce').fillna(df['review_rate_number'].median())
df['calculated_host_listings_count'] = pd.to_numeric(df['calculated_host_listings_count'], errors='coerce').fillna(df['calculated_host_listings_count'].median())
df['availability_365'] = pd.to_numeric(df['availability_365'], errors='coerce').fillna(df['availability_365'].median())

# Forward fill for last_review using ffill()
df['last_review'] = df['last_review'].ffill()

# Summary of missing values after handling
missing_values_summary_after = df.isna().sum()

# Display the summary of missing values after handling
missing_values_summary_after

id                                0
name                              0
host_id                           0
host_identity_verified            0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
lat                               0
long                              0
country                           0
country_code                      0
instant_bookable                  0
cancellation_policy               0
room_type                         0
construction_year                 0
price                             0
service_fee                       0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
review_rate_number                0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

> After applying these strategies, the dataset is now cleaner and more reliable for further analysis.

## Exploratory Data Analysis

In [30]:
# Display summary statistics
print("\nSummary Statistics:")
print(df.describe())


Summary Statistics:
                 id       host_id            lat           long  \
count  1.020500e+05  1.020500e+05  102050.000000  102050.000000   
mean   2.918655e+07  4.926846e+10      40.728097     -73.949652   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.509617e+07  2.459947e+10      40.688730     -73.982580   
50%    2.918659e+07  4.913046e+10      40.722290     -73.954440   
75%    4.327700e+07  7.400936e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   
std    1.627051e+07  2.853768e+10       0.055852       0.049502   

       construction_year          price    service_fee  minimum_nights  \
count           102050.0  102050.000000  102050.000000   102050.000000   
mean         2012.487026     625.361921     125.040568        8.094748   
min               2003.0      50.000000      10.000000    -1223.000000   
25%               2008.0     341.000000      68.000000        1.000000   
50%  

**Insights from Summary Statistics:**
- **Construction Year**: The average construction year is around 2012, with listings ranging from 2003 to 2022.
- **Price and Service Fee**: The average price is approximately `$625`, and the average service fee is around `$125`. Prices range from `$50` to `$1200`, and service fees range from `$10` to `$240`.
- **Minimum Nights**: The average minimum nights required is around 8, with some extreme outliers (e.g., -1223 and 5645).
- **Number of Reviews**: The average number of reviews per listing is about 27, with a maximum of 1024 reviews.
- **Last Review**: The dates of the last reviews range from 2012 to 2058, indicating some potential data entry errors.
- **Reviews per Month**: The average number of reviews per month is approximately 1.28, with a maximum of 90.
- **Review Rate Number**: The average review rating is around 3.28, with a range from 1 to 5.
- **Calculated Host Listings Count**: The average number of listings per host is about 7.92, with a maximum of 332.
- **Availability 365**: The average availability is around 141 days per year, with some extreme outliers (e.g., -10 and 3677).

These statistics provide a comprehensive overview of the dataset, highlighting key attributes and potential areas for further investigation.

#### Handling Abnormal Values
- **Outlier Detection**: Used the Interquartile Range (IQR) method to identify outliers in numerical columns.
- **Outlier Handling**: Replaced outliers with the median value of the respective column to ensure the dataset is more robust and reliable for analysis.

After handling the abnormal values, the dataset is now cleaner and ready for further exploratory data analysis (EDA).


In [31]:
# Function to identify and handle outliers using IQR
def handle_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Replace outliers with the median value
    median_value = df[column].median()
    df[column] = np.where((df[column] < lower_bound) | (df[column] > upper_bound), median_value, df[column])

# List of numerical features to check for outliers
numerical_features = ['price', 'service_fee', 'minimum_nights', 'number_of_reviews', 'reviews_per_month', 'review_rate_number', 'calculated_host_listings_count', 'availability_365']

# Apply the function to each numerical feature
for feature in numerical_features:
    handle_outliers(df, feature)

# Summary of the dataset after handling outliers
df.describe()

id       host_id            lat           long  \
count  1.020500e+05  1.020500e+05  102050.000000  102050.000000   
mean   2.918655e+07  4.926846e+10      40.728097     -73.949652   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.509617e+07  2.459947e+10      40.688730     -73.982580   
50%    2.918659e+07  4.913046e+10      40.722290     -73.954440   
75%    4.327700e+07  7.400936e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   
std    1.627051e+07  2.853768e+10       0.055852       0.049502   

       construction_year          price    service_fee  minimum_nights  \
count           102050.0  102050.000000  102050.000000   102050.000000   
mean         2012.487026     625.361921     125.040568        2.736629   
min               2003.0      50.000000      10.000000       -5.000000   
25%               2008.0     341.000000      68.000000        1.000000   
50%               2012.0     625.000000     125.000000        3.000000   
75%               2017.0     912.000000     182.000000        3.000000   
max               2022.0    1200.000000     240.000000       11.000000   
std             5.759781     331.276858      66.238296        1.688150   

       number_of_reviews                    last_review  reviews_per_month  \
count      102050.000000                         102050      102050.000000   
mean           12.392043  2019-05-10 03:49:28.121509120           0.948618   
min             0.000000            2012-07-11 00:00:00           0.010000   
25%             1.000000            2018-10-19 00:00:00           0.280000   
50%             7.000000            2019-06-12 00:00:00           0.740000   
75%            15.000000            2019-07-03 00:00:00           1.200000   
max            73.000000            2058-06-16 00:00:00           3.850000   
std            16.448274                            NaN           0.892539   

       review_rate_number  calculated_host_listings_count  availability_365  
count       102050.000000                   102050.000000     102050.000000  
mean             3.277923                        1.268996        140.809799  
min              1.000000                        1.000000        -10.000000  
25%              2.000000                        1.000000          3.000000  
50%              3.000000                        1.000000         96.000000  
75%              4.000000                        1.000000        268.000000  
max              5.000000                        3.000000        426.000000  
std              1.283152                        0.569855        134.713210

**Summary Statistics:**
- **Construction Year**: The average construction year is around 2012, with listings ranging from 2003 to 2022.
- **Price and Service Fee**: The average price is approximately `$625`, and the average service fee is around `$125`. Prices range from `$50` to `$1200`, and service fees range from `$10` to `$240`.
- **Minimum Nights**: The average minimum nights required is around 2.74, with some extreme outliers previously handled.
- **Number of Reviews**: The average number of reviews per listing is about 12.39, with a maximum of 73 reviews.
- **Last Review**: The dates of the last reviews range from 2012 to 2058, indicating some potential data entry errors.
- **Reviews per Month**: The average number of reviews per month is approximately 0.95, with a maximum of 3.85.
- **Review Rate Number**: The average review rating is around 3.28, with a range from 1 to 5.
- **Calculated Host Listings Count**: The average number of listings per host is about 1.27, with a maximum of 3.
- **Availability 365**: The average availability is around 141 days per year, with some extreme outliers previously handled.

#### Handling Abnormal Values in `last_review`
- **Date Conversion**: Converted `last_review` to datetime format.
- **Date Range**: Inferred the start date and end date from the dataset. Ensured the end date does not exceed `2024-08-31`.
- **Outlier Handling**: Replaced dates outside this range with NaN and filled NaN values with the median date.

After handling the abnormal values in `last_review`, the dataset is now cleaner and more reliable for further analysis.


In [32]:
# Convert last_review to datetime if not already done
df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')

# Infer start and end dates from the dataset
start_date = df['last_review'].min()
end_date = df['last_review'].max()

# Ensure the end date does not exceed 2024
end_date = min(end_date, pd.to_datetime('2024-08-31'))

print(f"Start Date: {start_date}")
print(f"End Date: {end_date}")

# Replace dates outside the range with NaN
df['last_review'] = df['last_review'].where(df['last_review'].between(start_date, end_date))

# Fill NaN values with the median date
median_date = df['last_review'].median()
df['last_review'].fillna(median_date, inplace=True)

# Summary of the dataset after handling abnormal values in last_review
df['last_review'].describe()

Start Date: 2012-07-11 00:00:00
End Date: 2024-08-31 00:00:00


count                           102050
mean     2019-05-09 21:33:58.083292672
min                2012-07-11 00:00:00
25%                2018-10-19 00:00:00
50%                2019-06-12 00:00:00
75%                2019-07-03 00:00:00
max                2024-08-15 00:00:00
Name: last_review, dtype: object

**Date Range**:
- **Start Date**: The earliest `last_review` date is `2012-07-11`.
- **End Date**: The latest `last_review` date is `2024-08-15`.

**Summary Statistics**:
- **Mean Date**: The average `last_review` date is around `2019-05-09`.
- **Median Date**: The median `last_review` date is `2019-06-12`.
- **Interquartile Range**:
  - 25th Percentile: `2018-10-19`
  - 75th Percentile: `2019-07-03`

These statistics indicate that the majority of reviews are relatively recent, with most falling between late 2018 and mid-2019. The dataset now has a consistent and reasonable range of dates for `last_review`, making it more reliable for further analysis.
